src/Microsoft.Health.Fhir.Core/Features/Operations/Reindex/ReindexJobTask.cs

```diff
@@ -126,96 +124,207 @@ public async Task ExecuteAsync(ReindexJobRecord reindexJobRecord, WeakETag weakE
+
+#pragma warning disable CS4014 // Suppressed as we want to continue execution and begin processing the next query while this continues to run
+                        queryTasks.Add(ProcessQueryAsync(query, jobSemaphore, queryTokensSource.Token));
+#pragma warning restore CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
 
-                        await ProcessQueryAsync(query, cancellationToken);
+                        _logger.LogInformation($"Reindex job task using {queryTasks.Count} threads");
+                    }
+
+                    // reset stale queries to pending
+                    var staleQueries = _reindexJobRecord.QueryList.Where(
+                        q => q.Status == OperationStatus.Running && q.LastModified < DateTimeOffset.UtcNow - _reindexJobConfiguration.JobHeartbeatTimeoutThreshold);
+                    foreach (var staleQuery in staleQueries)
+                    {
+                        await jobSemaphore.WaitAsync();
+                        try
+                        {
+                            // if this query has a created task, cancel it
+                            if (queryCancellationTokens.TryGetValue(staleQuery, out var tokenSource))
+                            {
+                                tokenSource.Cancel(false);
+                            }
+
+                            staleQuery.Status = OperationStatus.Queued;
+                            await UpdateJobAsync(cancellationToken);
+                        }
+                        finally
+                        {
+                            jobSemaphore.Release();
+                        }
+                    }
+
+                    await Task.Delay(_reindexJobConfiguration.QueryDelayIntervalInMilliseconds);
+
+                    // Remove all finished tasks from the collections of tasks
+                    // and cancellationTokens
+                    if (queryTasks.Count >= reindexJobRecord.MaximumConcurrency)
+                    {
+                        var taskArray = queryTasks.ToArray();
+                        Task.WaitAny(taskArray);
+                        var finishedTasks = queryTasks.Where(t => t.IsCompleted).ToArray();
+                        foreach (var finishedTask in finishedTasks)
+                        {
+                            queryTasks.Remove(finishedTask);
+                            queryCancellationTokens.Remove(finishedTask.Result);
+                        }
+                    }
+
+                    // if our received CancellationToken is cancelled we should
+                    // pass that cancellation request onto all the cancellationTokens
+                    // for the currently executing threads
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        foreach (var tokenSource in queryCancellationTokens.Values)
```

Comment:
The comment sounds like we should be using `CancellationTokenSource.CreateLinkedTokenSource` for the these